modules/core/src/main/java/org/apache/ignite/internal/client/impl/connection/GridClientNioTcpConnection.java

```diff
@@ -469,7 +469,18 @@ else if (now - lastPingSndTime > pingInterval && lastPingRcvTime != Long.MAX_VAL
 
             assert old == null;
 
-            GridNioFuture<?> sndFut = ses.send(msg);
+            GridNioFuture<?> sndFut;
+            if(sesTok == null && credentials() != null ) {
+                fut.retryState(TcpClientFuture.STATE_AUTH_RETRY);
+
+                GridClientAuthenticationRequest req = buildAuthRequest();
+
+                req.requestId(reqId);
+
+                sndFut = ses.send(req);
```

Comment:
I don't see where you send "msg" in this branch. Is it ignored?

Reply:
The message is already pended in the original future (for which status is changing to STATE_AUTH_RETRY). So the message is not ignored but postponed.